### PR TITLE
[GLT-3902] Project Assays

### PR DESCRIPTION
**Pull request by aryan2gandevia**
Jira ticket or GitHub issue: https://jira.oicr.on.ca/browse/GLT-3902

✅ Includes a change file

### CODE REVIEW

File: miso-web/src/main/webapp/scripts/form_requisition.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/form_requisition.js]: Remove

File: miso-web/src/main/webapp/scripts/form_requisition.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/form_requisition.js:31]: Can use `Array.find` instead of `filter` for finding a single item, but we also have a `Utils.array.findUniqueOrThrow` (`lims.js`) that works better here

File: miso-web/src/main/webapp/scripts/form_requisition.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/form_requisition.js:39]: Make sure it's not already there before adding again

File: miso-web/src/main/webapp/scripts/form_sample.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/form_sample.js:75]: Why was this moved?

File: miso-web/src/main/webapp/scripts/list_assay.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_assay.js]: * Why did this change from `isAdmin` to `admin`?
* The admin options should not be included on the project page

File: miso-web/src/main/webapp/scripts/list_assay.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_assay.js:38]: Admin options should not be included on the project page

File: miso-web/src/main/webapp/scripts/project_ajax.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/project_ajax.js:3]: Why are you concerned with `FormTarget` here?

File: sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java
Comment (djcooke) [inline sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java:10]: ?

File: sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java
Comment (djcooke) [inline sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java]: * could be reduced to `assertTrue(assays.size() > 0)`
* better to test specific value. There should be 2 assays, so `assertEquals(2, assays.size())`
* even better to test that the correct assays are added (assay IDs 1 and 2). Note you can't depend on these being returned in any particular order, so can do something like `assertTrue(assays.stream().anyMatch(...))`

File: miso-web/src/main/webapp/scripts/form_sample.js
Comment (aryan2gandevia) [reply miso-web/src/main/webapp/scripts/form_sample.js:74]: It was because in the bulk_sample.js file I moved the project section before the requisition field, so I thought I'd keep the order uniform. However, if there is copy-pasting happening, then I'll revert the change to prevent potential errors

File: sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java
Comment (aryan2gandevia) [reply sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java:74]: Commented out to test something during submission, will un-comment and make the proper test before merging.

File: docs/user_manual/projects.md
Comment (djcooke) [inline docs/user_manual/projects.md:94]: * remove all backticks (inconsistent with the rest of the document)
* describe _how_ to add and remove assays
* add a link to the Assays of the user manual (inside the Requisitions page) for more information about assays

File: changes/change_GLT-3902.md
Comment (djcooke) [inline changes/change_GLT-3902.md:1]:
```suggestion
Assay options on the Edit Requisition page are now limited to the assays assigned to all of the requisitioned samples' projects
```

File: miso-web/src/main/webapp/scripts/list_assay.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/list_assay.js:15]: `>= 0` isn't ideal to check if the value is a number here because in JavaScript `null >= 0` is true. `projectId` will usually be undefined, so this will work currently, but there's potential for a confusing bug if it's ever set to null. Change to just check for a truthy value instead (`if (config.projectId) {...`)

Same below for static actions

File: sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java
Comment (djcooke) [inline sqlstore/src/it/java/uk/ac/bbsrc/tgac/miso/persistence/impl/HibernateProjectDaoIT.java:83]: Check the size as well to ensure there are no extra unexpected assays

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/view/RequisitionController.java
Comment (aryan2gandevia) [reply miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/view/RequisitionController.java:174]: Yes, all assays are available to choose from when creating an assay.

I put it as an empty list in the model because if it isn't there, then calling it in editRequisition.jsp would cause an error. If, for example, I had:
`var holder = ${potentitalAssays};`
 and if I didn't add it to the model, then it would become
` var holder = ; `
 which causes an error. 
 
 I've tried using conditionals to catch this in different ways but haven't found a way without adding a bunch of extra code.

Comment (djcooke) [reply miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/controller/view/RequisitionController.java:174]: Only access it if the `pageMode == "edit"`

File: miso-web/src/main/webapp/scripts/bulk_sample.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/bulk_sample.js:682]: * duplicated code - move to a function
* array and for loop are unnecessary. Can reduce to

    ```
    source: projectSelected.assayIds.map(function(assayId) {
        return Utils.array.findUniqueOrThrow(Utils.array.idPredicate(assayId), Constants.assays);
    }),
    ```

File: miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultAssayService.java
Comment (aryan2gandevia) [inline miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultAssayService.java:1]: The changes in the file are just the develop branch's `DefaultAssayService.java` new code as it was changed and caused errors unrelated to what I a added when running tests.

File: miso-web/src/main/webapp/scripts/bulk_sample.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/bulk_sample.js:689]: * duplicate code - move to function
* filter out archived assays

File: miso-web/src/main/webapp/scripts/bulk_sample.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/bulk_sample.js:1591]: Misleading name as it doesn't return anything. `updateAssayOptions` maybe?

File: miso-web/src/main/webapp/WEB-INF/pages/editProject.jsp
Comment (djcooke) [inline miso-web/src/main/webapp/WEB-INF/pages/editProject.jsp:185]: This is not useful to the form, and the rest of the config is not relevant to the list. Instead of adding this to the form config, give the list its own config. Can be inline below: `Project.setListConfig({projectId: projectDto.id})`

File: miso-web/src/main/webapp/scripts/bulk_sample.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/bulk_sample.js:1603]: Don't clear the value, as the user might not notice this. If it become invalid with the change, it will appear red

**Review by djcooke — COMMENTED** [review]
It's not obvious why there are sometimes no assay options when bulk creating samples. Can you add a description to the assay column (in `bulk.js`) something like "The assay assigned to the requisition the sample belongs to. When receiving libraries, the options are limited to the assays assigned to the project." (this will appear in the expandable Quick Help section at the top of the page). Add a description to the Edit Requisition page to explain the options too

File: miso-web/src/main/webapp/scripts/bulk.js
Comment (djcooke) [inline miso-web/src/main/webapp/scripts/bulk.js:784]: My mistake, but should be "when receiving samples"

**Review by djcooke — APPROVED** [review]
Approved after the text fix I mentioned

**Review by callunity — APPROVED** [review]

**Pull request merged**
